ADALiOS/ADALiOS/ADTokenCache.m

```diff
@@ -0,0 +1,546 @@
+                                           error:(ADAuthenticationError *__autoreleasing *)error
+{
+    (void)error;
+    
+    @synchronized(self)
+    {
+        [_delegate willAccessCache:self];
+        NSArray<ADTokenCacheItem *> * result = [self getItemsImplKey:key userId:userId];
+        [_delegate didAccessCache:self];
+        
+        return result;
+    }
+}
+
+- (NSArray<ADTokenCacheItem *> *)getItemsImplKey:(nullable ADTokenCacheKey *)key
+                                          userId:(nullable NSString *)userId
+{
+    if (!_cache)
+    {
+        return nil;
+    }
+    
+    NSMutableArray* items = [NSMutableArray new];
+    NSDictionary* tokens = [_cache objectForKey:@"tokens"];
+    if (!tokens)
+    {
+        return nil;
+    }
+    
+    
+    NSDictionary* idtokens = [_cache objectForKey:@"idtokens"];
+    if (userId)
+    {
+        // If we have a specified userId then we only look for that one
+        [self addToItems:items forUserId:userId tokens:tokens idtokens:idtokens key:key];
+    }
+    else
+    {
+        // Otherwise we have to traverse all of the users in the cache
+        for (NSString* userId in tokens)
+        {
+            [self addToItems:items forUserId:userId tokens:tokens idtokens:idtokens key:key];
+        }
+    }
+    
+    return items;
+}
+
+
+/*! Clears token cache details for specific keys.
+ @param key: the key of the cache item. Key can be extracted from the ADTokenCacheItem using
+ the method 'extractKeyWithError'
+ @param userId: The user for which the item will be removed. Can be nil, in which case items for all users with
+ the specified key will be removed.
+ The method does not raise an error, if the item is not found.
+ */
+- (BOOL)removeItem:(ADTokenCacheItem *)item
+             error:(ADAuthenticationError * __autoreleasing *)error
+{
+    [_delegate willWriteCache:self];
+    BOOL result = [self removeImpl:item error:error];
+    [_delegate didWriteCache:self];
+    return result;
+}
+
+- (BOOL)removeImpl:(ADTokenCacheItem *)item
+             error:(ADAuthenticationError * __autoreleasing *)error
+{
+    ADTokenCacheKey* key = [item extractKey:error];
+    if (!key)
+    {
+        return NO;
+    }
+    
+    NSString* userId = item.userInformation.userId;
+    if (!userId)
+    {
+        userId = @"";
+    }
+    
+    NSMutableDictionary* tokens = [_cache objectForKey:@"tokens"];
+    if (!tokens)
+    {
+        return YES;
+    }
+    
+    NSMutableDictionary* userTokens = [tokens objectForKey:userId];
+    if (!userTokens)
+    {
+        return YES;
+    }
+    
+    if (![userTokens objectForKey:key])
+    {
+        return YES;
+    }
+    
+    [userTokens removeObjectForKey:key];
+    
+    // Check to see if we need to remove the idtoken
+    if (!userTokens.count)
+    {
+        [tokens removeObjectForKey:userId];
+        [[_cache objectForKey:@"idtokens"] removeObjectForKey:userId];
+    }
+    
```

Comment:
Consider adding logging statements for count and cache key. It will help make the logs read like functional specs.